### DISCUSSION TITLE
Federated Instances

### DISCUSSION THREAD

**Discussion by erlend-sh** [muni-town/weird#49] [General]
We know we want to support some form of federation between different instances of Weird, e.g. [weird.one](https://weird.one), [muni.town](https://muni.town/) etc., but how exactly we'll do that remains unclear.

Federation would make it possible for self-hosted instances, single-user and multi-user alike, to opt-in to Weird's global [Network of Shared Purpose](https://github.com/commune-os/weird/issues/3) without relying on a central authority.

However, it's possible that we could approach federation in stages. _Starting_ with a centralized approach is fine as long as there's no lock-in, and a fairly clear path towards true federation/decentralisation.

### Federation Lite

For example, https://stackexchange.com/ is a kind of 'federation lite', in that it supports a bunch of distinct spaces (including the top level domain stackoverflow.com) but they're all controlled by the same StackExchange account.

We could run a similar model by having `weird.one` be the (provisional) central authority that lets people tie their sites into the larger network. The way I could see that working is a lot like https://webmention.io/

- You set up your own instance of Weird on `mysite.com`
- You go to weird.one and sign in via IndieAuth or our custom equivalent of it
- Your site and its profile would be added to the Weird Network, maybe unlisted pending peer review

Or maybe it’s just an in-app OAuth authorization..
![IMG_2364](https://github.com/commune-os/weird/assets/583842/6addf11d-b22d-401a-b832-4c8086f74384)

<details>
  <summary>Discord Connections</summary>

![IMG_2363](https://github.com/commune-os/weird/assets/583842/48e14af5-5dc4-4ed0-9125-36a2bb550734)

</details>

A whole multi-user network like `ournetwork.com` could follow a similar process, but it'd be done on behalf of all users on a sitewide level, so anyone who signs up for `ournetwork.com` would automatically be part of the Weird Network. Individual opt-in models are also possible.

### Real Federation

We have several technologies at our disposal which could hopefully behave in complementary ways:

- https://openid.net/specs/openid-federation-1_0.html
  - https://www.authlete.com/developers/oidcfed/
- https://developers.google.com/privacy-sandbox/3pcd/fedcm
- [Iroh](https://iroh.computer/) & [Willow](https://willowprotocol.org/)

**Comment by zicklag** [comment]
The [Technical Details](https://zicklag.katharos.group/blog/how-to-federate/#technical-details) section of my [How to Federate?](https://zicklag.katharos.group/blog/how-to-federate/) post has an overview of some of the challenges and solutions for federating Weird.

Some of these solutions will actually fix scalability problems in the way we currently do things, like providing an index for usernames so we don't have to loop through every user account every time we display a profile.

I'm going to create issues for different steps involved in federation and add a `federation` label to them. I think it will make sense to tackle these things bit by bit, but we'll feel it out. It may be possible to get Weird and Muni town federating within a week or two, even while adding features like links, avatars, etc. that we want to get in.

**Comment by erlend-sh** [comment]
Just a couple park walk notes:

At its essence, the *Website Federation* we wanna do is according to conventions like `/now`

> ### What is a “now page”?
> 
> Most websites have a link that says “about”. It goes to a page that tells you something about the background of this person or business. For short, people just call it an “about page”.
> 
> Most websites have a link that says “contact”. It goes to a page that tells you how to contact this person or business. For short, people just call it a “contact page”.
> 
> So a website with a link that says “now” goes to a page that tells you what this person is focused on at this point in their life. For short, we call it a “now page”.
> 
> See examples by browsing [nownownow.com](https://nownownow.com/).
> 
> Although it’s normal to make the web address “/now”, just like it’s usually “/about” and “/contact” in those other examples, the URL could be anything.

We take those cultural conventions and we polish those organic designs with a UX pass of our own. Any bespoke website that follows these conventions should have a very easy time hooking up with the Weird network in order to show up in our directory of people-sites.

**Comment by aschrijver** [comment]
> but how exactly we'll do that remains unclear.

This thread is a bit older. Is my understanding correct that Leaf is the chosen way forward?

**Reply by zicklag** [reply]
Yes, Leaf is the plan for federation.

The idea is that we will use domains as usernames and resolve them to the user's Leaf Subspace ID ( public key ) and also a peer ID that has the user's data on it.

This will allow the main Weird instance to load user profiles that are on other Weird server instances, or even ones that are hosted using the ( not yet created ) Weird desktop app without servers.

**Reply by zicklag** [reply]
There are still some questions about how exactly we will will handle the global listing of users across different instances.

For example, while the above mentioned Leaf federation would allow you to view a user's profile by going to `weird.one/zicklag.example.com`, even if `zicklag.example.com`'s profile is on another server, it wouldn't necessarily put that user in the global user index on Weird.one.

We would probably require you to sign up for a free weird account if you wanted to list your profile in the global index, but your profile data could still be stored on another server.

That would help avoid bot spam and allow us to blacklist users for moderation purposes.

**Reply by aschrijver** [reply]
> This will allow the main Weird instance to load user profiles that are on other Weird server instances, or even ones that are hosted using the ( not yet created ) Weird desktop app without servers.

Some feedback on domain terminology. Weird / Leaf I think should consider being clearer on domain model and ubiquitous language. The sentence above in its choice of words - though technically accurate for current MVP scope - may give a wrong depiction of the domain.

Though it depends on your long-term plans, regarding proposition. Like choice between positioning e.g. as:

1. Weird distributed app framework with a community.
2. Leaf protocol with an open ecosystem, and Weird a participant.

Going from a perspective of 2) and revisiting the text, do we still want to refer to a "Weird instance" and to "server instance"? How much of the "local-first" paradigm can we capture to convey new concepts well?

"Leaf" itself is a very appropriate term. A server implementation of Leaf protocol can be simply indicated as a Leaf. If we go Martin Kleppmann's direction then Leaf protocol offers generic sync and Leaf nodes become like universal back-ends. What does that make Weird in an ecosystem with many other products and services? How are these consumed?

**Reply by zicklag** [reply]
The positioning of weird is as in option 2: "Leaf protocol with an open ecosystem, and Weird a participant."

As far as Leaf is concerned we don't have a concept of "server instances". Every participant is just a Leaf node.

In the case of Weird, though, we are, in a way, bridging the p2p / local-first and Web2 worlds, and doing that requires server instances.

In Weird we have instances as a way create an online presence and a typical Web2 onboarding process, while building on Leaf and starting to allow all of the centralized server stuff to be optional.

Once we start getting more documentation together, a proper glossary of terms can be developed.

**Reply by erlend-sh** [reply]
> Like choice between positioning

I don’t think we need to choose. Weird docs will explain things from Weird’s POV, and Leaf docs will explain things from Leaf’s POV.

We can hold both simultaneously. The tools and networks we’re making will mean different things to different people, depending on how they’re engaging with our stuff.

Leaf protocol is for the web-native computer nerds.
Weird app is for all internet nerds.

> If we go Martin Kleppmann's direction then Leaf protocol offers generic sync and Leaf nodes become like universal back-ends. What does that make Weird in an ecosystem with many other products and services? How are these consumed?

From Leaf’s POV, Weird is a consumer of Leaf, i.e. an implementation of the Leaf protocol.

From Weird’s POV, Leaf is a cornerstone of the app, but still just one stone of many components (Peaf, Rauthy, Svelte, Tauri (soon)) that make up Weird as a platform and a *place* (a network-destination).

**Reply by zicklag** [reply]
> I don’t think we need to choose. 

That's a good point. :+1: 

Especially when some of our future plans for Weird include a plugin-based CMS-type thing, which will itself become a kind of "platform".

Weird and Leaf are just different layers, and somebody who's interested in local first app development might go straight to Leaf, and somebody who wants a website will go straight to Weird.

**Reply by aschrijver** [reply]
While that is a workable description, it contains a number of assumptions that should be clarified further. From the text I might conclude that things are very similar to the current Fediverse. You have Apps like Weird and a protocol Leaf for app federation and potentially to interoperate with other apps. Apps are otherwise independent platforms spun up by their server instances.

> From Weird’s POV, Leaf is a cornerstone of the app, but still just one stone of many components (Peaf, Rauthy, Svelte, Tauri (soon)) that make up Weird as a platform and a _place_ (a network-destination).

If things are indeed like current Fediverse, there isn't something that is meaningfully a Place. There are app siloes. Apps typically first federate with themselves between instances, and in later project stage start to consider with whom to interoperate, and then deliver thin and fragile 'pipes' of integration doing whack-a-mole programming and maintenance.

It is a very technical perspective to look at the network. Weird its own independent tech stack offering a platform (that may be extended further) and 3rd parties building their own platforms, sprinkled loosely with 'interop bridges' between them.

That is not how you see Weird/Leaf I am sure. Just to sketch some thoughts, brainstorming:

- The conceptual model of the "agentic fediverse" should be the basis for the DX of ecosystem participants.
  - In current fedi conceptually we have "actors exchanging activities on their social graph", yet in practice stuff the DX is stuck on a much lower technical level.
- In Leaf protocol you describe what is _inside_ Leaf. But Leaf itself can be an interesting concept to include in the model.
  - A Leaf is a local-first data store that syncs in the background. DX-wise there is no network anymore. All state + data is local.
  - A Leaf topology then has very specific characteristics, that can be worth elaborating further.
  - A Leaf Node could be name of a technical component to be integrated in a web app or a local-first app (e.g. Tauri) or run stand-alone.
- How would a "Place" conceptually look like? You mentioned "network destination".
  - So I imagine an empty social landscape, and with Weird a bustling communal town emerges. But is a 3rd-party app then a rival town?
  - Place indicates locality, and that should maybe be reserved for real geographic areas.
  - There may be good real-world analogies that can inspire the model e.g. with landscape, nature, plant life.

PS. You consider Weird a special case as a permanent running peer, kind of centralized component, while still allowing p2p on the side. I think that shouldn't be special case, but fit in the model.

- In p2p it is very common there are [super nodes](https://en.wikipedia.org/wiki/Supernode_(networking)).
- Any Leaf can be a regular peer (client) or a superpeer (relay). And Weird just hosts its Leaf Node as a superpeer.

(Pff, I'll stop as once again I seem to have difficulty expressing myself well. Hope the gist makes sense. Wanted to follow separarating technical and conceptual perspectives and list some possible terminology, but yea ..)

Technical perspective:

- Leaf Protocol: Open specification for peer-to-peer local-first data synchronisation.
- Leaf SDK: Libraries and tools
- Leaf Server: Reference implementation of a self-hostable Leaf Host.

**Reply by erlend-sh** [reply]
> If things are indeed like current Fediverse

They’re not. There are similarities, but Leaf-net is a different type of network.

It’s a network of federated instances, yes. But any other preconceptions of what ‘the fediverse’ is are best put aside if you want to learn about Weird/Leaf with a Beginner’s Mind, which will be the target audience of our marketing and educational materials.

> How would a "Place" conceptually look like? You mentioned "network destination".

This is gonna be a lot easier to show rather than tell, but in short, I think of every subreddit, Discord server and Wikipedia entry as a *place* on the web, because there are people congregating there.

> I imagine an empty social landscape, and with Weird a bustling communal town emerges. But is a 3rd-party app then a rival town?

On one level, yes. But only insofar as all neighboring towns in our world are ‘rivals’ to some extent, in the game of competing for people’s attention/occupancy.

But in a healthy nation/state/municipality/network, all the towns are chiefly playing cooperative games together, because they’re part of the same ecosystem, and there’s enough good will and economical activity going around for everyone.

> Hope the gist makes sense.

I think we’ve got some good sense-making in progress here, yeah! 😁

**Reply by zicklag** [reply]
> PS. You consider Weird a special case as a permanent running peer, kind of centralized component, while still allowing p2p on the side. I think that shouldn't be special case, but fit in the model.

Yes, for now it ends up something that "fits into the model" fine, but is also a "special case", simply because we don't know for sure what the most general look at a Leaf "supernode" / instance will be.

At this point we have a pretty good idea of what we want Leaf to be like, at least in it's initial form, but it still leaves a lot unspecified, such as how do you deal with web2 identities, etc.

Weird currently represents a specific, experimental way of dealing with that, and there are multiple different ways you could bring the value of Leaf's purely p2p model, into a more approachable web-centric user-experience.

Until Weird is more fleshed out, and likely until there are more apps similar to Weird with their own experiments, it will be harder to see what a "general supernode pattern" would look like for Leaf, and that's OK.

Things are in layers so that we can more confidently say, "we're happy with how this is working, up to layer x. Everything after that will need some time to experiment and find good patterns."

---

PS: I'd be happy to hop on a call with you sometime if you wanted to ask more questions "in person". We've got a lot of stuff to document in the coming months, but any kind of preparatory discussion on what you're looking for on this kinds of stuff could be useful.
